File: packages/evals/tasks/bench/extract/iframe_hn.ts
import { z } from "zod";
import { defineBenchTask } from "../../../framework/defineTask.js";

export default defineBenchTask(
  { name: "iframe_hn" },
  async ({ debugUrl, sessionUrl, stagehand, page, logger }) => {
    try {
      await page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/iframe-hn/");

      // NOTE: the target content lives inside an iframe, but the task relies
      // entirely on extract() to see into it (no frame API usage in v3 or
      // v4), so this ported 1:1 — iframe handling is the SDK's responsibility.
      const { data: result } = await stagehand.extract(
        "extract the title of the first hackernews story",
        z.object({
          story_title: z.string(),
        }),
      );

      const title = result.story_title.toLowerCase();
      const expectedTitleSubstring = "overengineered anchor links";

      if (!title.includes(expectedTitleSubstring)) {
        logger.error({
          message: `Extracted title: ${title} does not contain expected substring: ${expectedTitleSubstring}`,
          level: 0,
        });
        return {
          _success: false,
          error: `Extracted title: ${title} does not contain expected substring: ${expectedTitleSubstring}`,
          logs: logger.getLogs(),
          debugUrl,
          sessionUrl,
        };
      }

      return {
        _success: true,
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    } catch (error) {
      return {
        _success: false,
        error: error instanceof Error ? error.message : String(error),
        logs: logger.getLogs(),
        debugUrl,
        sessionUrl,
      };
    } finally {
      await stagehand.close();
    }
  },
);

File: packages/sdk-ruby/evals/tasks/extract/iframe_hn.rb
# frozen_string_literal: true

# Port of packages/evals/tasks/bench/extract/iframe_hn.ts

Evals.define_task("iframe_hn") do |t|
  t.page.goto("https://browserbase.github.io/stagehand-eval-sites/sites/iframe-hn/")

  # NOTE: the target content lives inside an iframe, but the task relies
  # entirely on extract() to see into it (no frame API usage in v3 or
  # v4), so this ported 1:1 — iframe handling is the SDK's responsibility.
  result = t.stagehand.extract(
    "extract the title of the first hackernews story",
    schema: {
      "type" => "object",
      "properties" => { "story_title" => { "type" => "string" } },
      "required" => ["story_title"],
      "additionalProperties" => false,
    },
  )

  title = result.data["story_title"].downcase
  expected_title_substring = "overengineered anchor links"

  unless title.include?(expected_title_substring)
    t.logger.error("Extracted title: #{title} does not contain expected substring: #{expected_title_substring}")
    next {
      _success: false,
      error: "Extracted title: #{title} does not contain expected substring: #{expected_title_substring}",
    }
  end

  { _success: true }
end
